src/shaders/pbr.frag

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// This is an implementation of
+// Karis, Brian. “Real Shading in Unreal Engine 4 by.” (2013).
+
+// -------------- input ---------------------
+// position, normal, tangent in camera space
+in highp vec3 position;
+in highp vec3 normal;
+#if defined(TEXTURED)
+in highp vec2 texCoord;
+#endif
+#if defined(NORMAL_TEXTURE) && defined(PRECOMPUTED_TANGENT)
+in highp vec3 tangent;
+in highp vec3 biTangent;
+#endif
+
+// -------------- output -------------------
+layout(location = OUTPUT_ATTRIBUTE_LOCATION_COLOR) out vec4 fragmentColor;
+#if defined(OBJECT_ID)
+layout(location = OUTPUT_ATTRIBUTE_LOCATION_OBJECT_ID) out highp uint
+    fragmentObjectId;
+#endif
+
+#if (LIGHT_COUNT > 0)
+// -------------- lights -------------------
+// NOTE: In this shader, the light intensity is considered in the lightColor!!
+uniform vec3 LightColors[LIGHT_COUNT];
+uniform float LightRanges[LIGHT_COUNT];
+
+// if .w == 0, it means it is a dirctional light, .xyz is the direction;
+// if .w == 1, it means it is a point light, .xyz is the light position;
+// it is NOT put in the Light Structure, simply because we may modify the code
+// so it is computed in the vertex shader.
+uniform vec4 LightDirections[LIGHT_COUNT];
+#endif
+
+// -------------- material, textures ------------------
+struct MaterialData {
+  vec4 baseColor;     // diffuse color, if baseColorTexture exists,
+                      // multiply it with the baseColorTexture
+  float roughness;    // roughness of a surface, if roughness texture exists,
+                      // multiply it with the roughnessTexture
+  float metallic;     // metalness of a surface, if metallic texture exists,
+                      // multiply it the metallicTexture
+  vec3 emissiveColor; // emissiveColor, if emissive texture exists,
+                      // multiply it the emissiveTexture
+};
+uniform MaterialData Material;
+
+#if defined(BASECOLOR_TEXTURE)
+uniform sampler2D baseColorTexture;
+#endif
+#if defined(ROUGHNESS_TEXTURE)
+uniform sampler2D roughnessTexture;
+#endif
+#if defined(METALLIC_TEXTURE)
+uniform sampler2D metallicTexture;
+#endif
+#if defined(NORMAL_TEXTURE)
+uniform sampler2D normalTexture;
+#endif
+
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+uniform sampler2D packedTexture;
+#endif
+
+#if defined(EMISSIVE_TEXTURE)
+uniform sampler2D emissiveTexture;
+#endif
+
+// -------------- uniforms ----------------
+#if defined(OBJECT_ID)
+uniform highp uint objectId;
+#endif
+
+#if defined(NORMAL_TEXTURE) && defined(NORMAL_TEXTURE_SCALE)
+uniform mediump float normalTextureScale
+#ifndef GL_ES
+    = 1.0
+#endif
+    ;
+#endif
+
+// -------------- shader ------------------
+#if defined(NORMAL_TEXTURE)
```

Comment:
Does GLSL support #include's ? It seems this normal-mapping logic could be reused in other shaders. It's not specific to PBR.

Reply:
Not really, unless the shader is preprocessed externally first (which is what I'm working on right now, actually). However you can supply multiple source files and do it that way instead.

On the other hand, I don't think it's likely Habitat will have more shaders? As in, the shader will eventually be replaced with one from Magnum instead.

Reply:
I remembered I thought about this question and googled it a couple of years ago. And I decided not to share the code among shaders.
Take a look:
https://computergraphics.stackexchange.com/questions/100/sharing-code-between-multiple-glsl-shaders

Reply:
As magnum handles `#version` itself because of exactly the reasons mentioned in the SO answer, the only remaining valid point from there is

> 2. It's harder to control what needs to be included from within the shader (you need a separate system for this.)

Yes, that's true, but then again you have to prepend all the defines for `LIGHT_COUNT`, `NORMAL_TEXTURE` etc anyway, so having a separate external file doesn't make it *that much* worse.

Reply:
@mosra Adopting libshader_rc for Magnum would automatically handle the #INCLUDE preprocessing as well. :)

Reply:
@Skylion007 Yeah no :P Includes are supported directly by glslang itself, it's just not exposed in the `glslangValidator` command-line tool for whatever reason.

Which means, at this point, I don't see any benefits of using shaderc over glslang + spirv-tools directly, it would only add another hard-to-build dependency to the pipeline and further expand the surface for potential bugs.